File: core/src/rustyice/editor/EditorGameView.java
package rustyice.editor;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.scenes.scene2d.Event;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.FocusListener;
import rustyice.game.Game;
import rustyice.game.actors.Actor;
import rustyice.game.tiles.TileMap;
import rustyice.graphics.Camera;
import rustyice.graphics.GameDisplay;
import rustyice.graphics.GraphicsUtils;
import rustyice.graphics.RenderFlags;

public class EditorGameView {
    private EditorSelectionPane selectionPane;
    private EditorPropertyPane propertyPane;
    private GameDisplay display;

    private EditorInput editorInput;
    private Camera camera;
    private Game game;

    private boolean mouseDraging = false;
    private Vector2 mouseDownPos;
    private Vector2 mouseDraggedPos;
    private Actor selectionActor;
    private Vector2 selectionPos;

    private Vector2 tileSelectStart;
    private Vector2 tileSelectEnd;

    public EditorGameView(Game game){
        this.game = game;

        camera = new Camera();
        camera.setWidth(20);
        camera.setHeight(20);
        camera.enableFlag(RenderFlags.EDITOR);

        mouseDownPos = new Vector2();
        mouseDraggedPos = new Vector2();
        selectionActor = null;
        selectionPos = new Vector2();

        tileSelectStart = new Vector2();
        tileSelectEnd = new Vector2();

        display = new GameDisplay();
        display.setTarget(game, camera);

        editorInput = new EditorInput();
        display.addListener(editorInput);
    }

    public void render(SpriteBatch batch, float delta) {
        game.getCurrentSection().finishAdding();

        if (editorInput.moveUp) {
            camera.setY(camera.getY() + camera.getHeight() * delta);
        }
        if (editorInput.moveDown) {
            camera.setY(camera.getY() - camera.getHeight() * delta);
        }
        if (editorInput.moveLeft) {
            camera.setX(camera.getX() - camera.getWidth() * delta);
        }
        if (editorInput.moveRight) {
            camera.setX(camera.getX() + camera.getWidth() * delta);
        }

        if (editorInput.zoomIn) {
            camera.setWidth(camera.getWidth() * (1 - 0.75f * delta));
            camera.setHeight(camera.getHeight() * (1 - 0.75f * delta));
        }
        if (editorInput.zoomOut) {
            camera.setWidth(camera.getWidth() * (1 + 0.75f * delta));
            camera.setHeight(camera.getHeight() * (1 + 0.75f * delta));
        }

        if (Gdx.input.isKeyJustPressed(Input.Keys.L)) {
            camera.toggleFlag(RenderFlags.LIGHTING);
        }

        if(Gdx.input.isKeyJustPressed(Input.Keys.P)){
            camera.toggleFlag(RenderFlags.POV);
        }

        display.render(batch, delta);

        if (selectionPane.isTileMode() && mouseDraging) {
            display.getFBO().begin();
            batch.begin();

            batch.setColor(Color.LIGHT_GRAY);
            GraphicsUtils.drawRect(
                    tileSelectStart.x * TileMap.TILE_SIZE,
                    tileSelectStart.y  * TileMap.TILE_SIZE,
                    (tileSelectEnd.x - tileSelectStart.x)  * TileMap.TILE_SIZE,
                    (tileSelectEnd.y - tileSelectStart.y)  * TileMap.TILE_SIZE, 0.1f
            );

            batch.end();
            display.getFBO().end();
        }
    }

    public GameDisplay getDisplay(){
        return display;
    }

    public void dispose(){
        display.dispose();
    }

    public void setSelectionPane(EditorSelectionPane selectionPane){
        this.selectionPane = selectionPane;
    }

    public void setPropertyPane(EditorPropertyPane propertyPane){
        this.propertyPane = propertyPane;
    }

    private class EditorInput extends ClickListener {
        boolean moveUp, moveDown, moveLeft, moveRight, zoomIn, zoomOut;


        @Override
        public boolean keyDown(InputEvent event, int keycode) {
            switch (keycode){
                case Input.Keys.UP:
                    moveUp = true;
                    return true;
                case Input.Keys.DOWN:
                    moveDown = true;
                    return true;
                case Input.Keys.LEFT:
                    moveLeft = true;
                    return true;
                case Input.Keys.RIGHT:
                    moveRight = true;
                    return true;
                case Input.Keys.Z:
                    zoomIn = true;
                    return true;
                case Input.Keys.X:
                    zoomOut = true;
                    return true;
                default:
                    return false;
            }
        }

        @Override
        public boolean keyUp(InputEvent event, int keycode) {
            switch (keycode) {
                case Input.Keys.UP:
                    moveUp = false;
                    return true;
                case Input.Keys.DOWN:
                    moveDown = false;
                    return true;
                case Input.Keys.LEFT:
                    moveLeft = false;
                    return true;
                case Input.Keys.RIGHT:
                    moveRight = false;
                    return true;
                case Input.Keys.Z:
                    zoomIn = false;
                    return true;
                case Input.Keys.X:
                    zoomOut = false;
                    return true;
                default:
                    return false;
            }
        }

        @Override
        public void clicked(InputEvent event, float x, float y) {
            super.clicked(event, x, y);
            event.getStage().setKeyboardFocus(event.getListenerActor());

            if (selectionPane.isActorMode() && selectionPane.hasSelectedActor()) {
                Actor actor = selectionPane.buildSelectedActor();
                actor.setPosition(mouseDownPos.x, mouseDownPos.y);
                game.getCurrentSection().addActor(actor);
            }
        }

        @Override
        public boolean touchDown(InputEvent event, float x, float y, int pointer, int button) {
            super.touchDown(event, x, y, pointer, button);

            mouseDownPos.set(x, y);
            display.stageToSection(mouseDownPos);

            if (selectionPane.isActorMode()) {
                for (Actor actor : game.getCurrentSection().getActors()) {
                    if (mouseDownPos.x >= actor.getX() - actor.getWidth() / 2 && mouseDownPos.y >= actor.getY() - actor.getHeight() / 2
                            && mouseDownPos.x < actor.getX() + actor.getWidth() / 2 && mouseDownPos.y < actor.getY() + actor.getHeight() / 2) {
                        selectionActor = actor;
                        selectionPos.set(actor.getX(), actor.getY());
                        propertyPane.setSelected(selectionActor);
                        break;
                    }
                }
            }

            return true;
        }

        @Override
        public void touchUp(InputEvent event, float x, float y, int pointer, int button) {
            super.touchUp(event, x, y, pointer, button);
            selectionActor = null;

            if (selectionPane.hasSelectedTile() && mouseDraging) {
                for (int i = (int) tileSelectStart.y; i < tileSelectEnd.y; i++) {
                    for (int j = (int) tileSelectStart.x; j < tileSelectEnd.x; j++) {
                        game.getTiles().setTile(selectionPane.buildSelectedTile(), j, i);
                    }
                }
            }

            mouseDraging = false;
        }

        @Override
        public void touchDragged(InputEvent event, float x, float y, int pointer) {
            super.touchDragged(event, x, y, pointer);

            mouseDraging = true;
            mouseDraggedPos.set(x, y);
            display.stageToSection(mouseDraggedPos);

            if (selectionPane.isTileMode()) {
                if (mouseDownPos.x < mouseDraggedPos.x) {
                    tileSelectStart.x = tileFloor(mouseDownPos.x);
                    tileSelectEnd.x = tileCeil(mouseDraggedPos.x);
                } else {
                    tileSelectStart.x = tileFloor(mouseDraggedPos.x);
                    tileSelectEnd.x = tileCeil(mouseDownPos.x);
                }

                if (mouseDownPos.y < mouseDraggedPos.y) {
                    tileSelectStart.y = tileFloor(mouseDownPos.y);
                    tileSelectEnd.y = tileCeil(mouseDraggedPos.y);
                } else {
                    tileSelectStart.y = tileFloor(mouseDraggedPos.y);
                    tileSelectEnd.y = tileCeil(mouseDownPos.y);
                }
            }

            if (selectionActor != null) {
                selectionActor.setX(selectionPos.x + (mouseDraggedPos.x - mouseDownPos.x));
                selectionActor.setY(selectionPos.y + (mouseDraggedPos.y - mouseDownPos.y));
            }
        }

        private float tileFloor(float value){
            return MathUtils.floor(value / TileMap.TILE_SIZE);
        }

        private float tileCeil(float value){
            return MathUtils.ceil(value / TileMap.TILE_SIZE);
        }
    }
}

File: core/src/rustyice/editor/EditorGameView.kt
package rustyice.editor

import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.github.salomonbrys.kodein.instance
import rustyengine.RustyEngine
import rustyice.game.Actor
import rustyice.game.Game
import rustyice.game.tiles.TILE_SIZE
import rustyice.graphics.Camera
import rustyice.graphics.GameDisplay
import rustyice.graphics.drawRect

class EditorGameView(val game: Game, val selectionPane: EditorSelectionPane, val propertyPane: EditorPropertyPane){
    val display: GameDisplay

    private val editorInput: EditorInput
    private val camera: Camera

    private var mouseDraging: Boolean = false
    private val mouseDownPos: Vector2 = Vector2()
    private val mouseDraggedPos: Vector2 = Vector2()
    private var selectionActor: Actor? = null
    private val selectionDownPos: Vector2 = Vector2()

    private val tileSelectStart: Vector2 = Vector2()
    private val tileSelectEnd: Vector2 = Vector2()

    init {
        camera = Camera()
        camera.width = 20f
        camera.height = 20f

        display = GameDisplay()
        display.game = game
        display.camera = camera

        editorInput = EditorInput()
        display.addListener(editorInput)
    }

    fun render(batch: Batch, delta: Float) {
        game.currentSection?.finishAddingActors()

        if (editorInput.moveUp) {
            camera.y += camera.height * delta
        }
        if (editorInput.moveDown) {
            camera.y -= camera.height * delta
        }
        if (editorInput.moveLeft) {
            camera.x -= camera.width * delta
        }
        if (editorInput.moveRight) {
            camera.x += camera.width * delta
        }

        if (editorInput.zoomIn) {
            camera.width *= (1 - 0.75f * delta)
            camera.height *= (1 - 0.75f * delta)
        }
        if (editorInput.zoomOut) {
            camera.width *= (1 + 0.75f * delta)
            camera.height *= (1 + 0.75f * delta)
        }

        display.render(batch)

        if (selectionPane.isTileMode() && mouseDraging) {
            display.fbo!!.begin()
            batch.begin()

            batch.color = Color.LIGHT_GRAY
            batch.drawRect(
                    tileSelectStart.x * TILE_SIZE,
                    tileSelectStart.y  * TILE_SIZE,
                    (tileSelectEnd.x - tileSelectStart.x)  * TILE_SIZE,
                    (tileSelectEnd.y - tileSelectStart.y)  * TILE_SIZE, 0.1f
            )

            batch.end()
            display.fbo!!.end()
        }
    }

    fun dispose(){
        display.dispose()
    }

    private inner class EditorInput: ClickListener(){
        var moveUp = false
        var moveDown = false
        var moveLeft = false
        var moveRight = false
        var zoomIn = false
        var zoomOut = false


        override fun keyDown(event: InputEvent, keycode: Int):Boolean {
            when(keycode){
                Input.Keys.UP -> moveUp = true
                Input.Keys.DOWN -> moveDown = true
                Input.Keys.LEFT -> moveLeft = true
                Input.Keys.RIGHT -> moveRight = true
                Input.Keys.Z -> zoomIn = true
                Input.Keys.X -> zoomOut = true
                Input.Keys.R -> {
                        selectionActor?.let{ game.currentSection?.actors?.remove(it) }
                        selectionActor?.store()
                        selectionActor = null
                }
                else -> return false
            }
            return true
        }

        override fun keyUp(event: InputEvent, keycode: Int): Boolean {
            when (keycode) {
                Input.Keys.UP -> moveUp = false
                Input.Keys.DOWN -> moveDown = false
                Input.Keys.LEFT -> moveLeft = false
                Input.Keys.RIGHT -> moveRight = false
                Input.Keys.Z -> zoomIn = false
                Input.Keys.X -> zoomOut = false
                else -> return false
            }
            return true
        }

        override fun clicked(event: InputEvent, x: Float, y: Float) {
            super.clicked(event, x, y)
            event.stage.keyboardFocus = event.listenerActor

            if (selectionPane.isActorMode() && selectionPane.hasSelectedActor()) {
                val actor = selectionPane.buildSelectedActor()
                actor.setPosition(mouseDownPos.x, mouseDownPos.y)
                game.currentSection?.addActor(actor)
            }
        }

        override fun touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Boolean {
            super.touchDown(event, x, y, pointer, button)

            mouseDownPos.set(x, y)
            display.stageToSection(mouseDownPos)

            val section = game.currentSection
            if(section != null) {
                if (selectionPane.isActorMode()) {
                    for (actor in section.actors) {
                        if (mouseDownPos.x >= actor.x - actor.width / 2 && mouseDownPos.y >= actor.y - actor.height / 2
                                && mouseDownPos.x < actor.x + actor.width / 2 && mouseDownPos.y < actor.y + actor.height / 2) {
                            selectionActor = actor
                            selectionDownPos.set(actor.x, actor.y)
                            propertyPane.setSelected(actor)
                            break
                        }
                    }
                } else {
                    val tile = section.tiles.getTileAt(mouseDownPos.x, mouseDownPos.y)
                    if (tile != null) {
                        propertyPane.setSelected(tile)
                    }
                }
            }

            return true
        }

        override fun touchUp(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int) {
            super.touchUp(event, x, y, pointer, button)
            selectionActor = null

            val section = game.currentSection

            if (section != null && selectionPane.hasSelectedTile() && mouseDraging) {
                for (i in tileSelectStart.y.toInt() until tileSelectEnd.y.toInt()) {
                    for (j in tileSelectStart.x.toInt() until tileSelectEnd.x.toInt()) {
                        section.tiles.setTile(selectionPane.buildSelectedTile(), j, i)
                    }
                }
            }

            mouseDraging = false
        }

        override fun touchDragged(event: InputEvent, x: Float, y: Float, pointer: Int) {
            super.touchDragged(event, x, y, pointer)

            mouseDraging = true
            mouseDraggedPos.set(x, y)
            display.stageToSection(mouseDraggedPos)

            if (selectionPane.isTileMode()) {
                if (mouseDownPos.x < mouseDraggedPos.x) {
                    tileSelectStart.x = tileFloor(mouseDownPos.x)
                    tileSelectEnd.x = tileCeil(mouseDraggedPos.x)
                } else {
                    tileSelectStart.x = tileFloor(mouseDraggedPos.x)
                    tileSelectEnd.x = tileCeil(mouseDownPos.x)
                }

                if (mouseDownPos.y < mouseDraggedPos.y) {
                    tileSelectStart.y = tileFloor(mouseDownPos.y)
                    tileSelectEnd.y = tileCeil(mouseDraggedPos.y)
                } else {
                    tileSelectStart.y = tileFloor(mouseDraggedPos.y)
                    tileSelectEnd.y = tileCeil(mouseDownPos.y)
                }
            }

            selectionActor?.let {
                it.x = selectionDownPos.x + mouseDraggedPos.x - mouseDownPos.x
                it.y = selectionDownPos.y + mouseDraggedPos.y - mouseDownPos.y
            }
        }

        private fun tileFloor(value: Float): Float{
            return MathUtils.floor(value / TILE_SIZE).toFloat()
        }

        private fun tileCeil(value: Float): Float{
            return MathUtils.ceil(value / TILE_SIZE).toFloat()
        }
    }
}
